awx/main/scheduler/task_manager.py

```diff
@@ -483,24 +512,23 @@ def process_pending_tasks(self, pending_tasks):
                     found_acceptable_queue = True
                     break
 
-                if idle_instance_that_fits is None:
-                    idle_instance_that_fits = rampart_group.find_largest_idle_instance()
                 remaining_capacity = self.get_remaining_capacity(rampart_group.name)
                 if not rampart_group.is_containerized and self.get_remaining_capacity(rampart_group.name) <= 0:
                     logger.debug("Skipping group {}, remaining_capacity {} <= 0".format(
                                  rampart_group.name, remaining_capacity))
                     continue
 
-                execution_instance = rampart_group.fit_task_to_most_remaining_capacity_instance(task)
-                if execution_instance:
-                    logger.debug("Starting {} in group {} instance {} (remaining_capacity={})".format(
-                                 task.log_format, rampart_group.name, execution_instance.hostname, remaining_capacity))
-                elif not execution_instance and idle_instance_that_fits:
+                execution_instance = InstanceGroup.fit_task_to_most_remaining_capacity_instance(task, self.graph[rampart_group.name]['instances']) or \
```

Comment:
Conspiracy theory time, come along and join me @chrismeyersfsu @fosterseth 

1.  The task manager starts, obtains the lock, calls `after_lock_init`, and has a set of known instances.  We arrive at a moment in time _right before_ this code.
2. The user removes an instance via the API, or via `awx-manage deprovision_instance`, and it's the one we happen to pick here.
3.  What _terrible_ thing happens next?

Reply:
hmm good point. Not sure of the behavior, but this method seems to go along with this scenario https://github.com/ansible/awx/blob/e6c124962bca6215b8b6d4673ec252195df234be/awx/main/scheduler/task_manager.py#L533

Reply:
I think this is *very unlikely* to happen, but we should probably test to make sure that it doesn't result in something bad, like a task never running, and getting stuck in `waiting` forever (or worse, a task-manager breaking exception).

Reply:
okay I tried this on a local 3 node cluster install. If I delete a node after `execution_instance` is set, but the task has not yet saved, the job will fail but then I get these errors on an unending loop

```
tools_awx_1_1 | 2020-10-12 18:47:13,473 ERROR    awx.main.wsbroadcast AWX is currently installing/upgrading.  Trying again in 5s...
.
.
tools_awx_1_1 | 2020-10-12 18:47:16,633 ERROR    awx.main.dispatch encountered an error communicating with redis to store dispatcher statistics
tools_awx_1_1 | Traceback (most recent call last):
tools_awx_1_1 |   File "/awx_devel/awx/main/dispatch/worker/base.py", line 111, in record_statistics
tools_awx_1_1 |     self.redis.set(f'awx_{self.name}_statistics', self.pool.debug())
tools_awx_1_1 |   File "/awx_devel/awx/main/dispatch/pool.py", line 341, in debug
tools_awx_1_1 |     self.cleanup()
tools_awx_1_1 |   File "/awx_devel/awx/main/dispatch/pool.py", line 436, in cleanup
tools_awx_1_1 |     reaper.reap(excluded_uuids=running_uuids)
tools_awx_1_1 |   File "/awx_devel/awx/main/dispatch/reaper.py", line 38, in reap
tools_awx_1_1 |     (changed, me) = Instance.objects.get_or_register()
tools_awx_1_1 |   File "/awx_devel/awx/main/managers.py", line 150, in get_or_register
tools_awx_1_1 |     return (False, self.me())
tools_awx_1_1 |   File "/awx_devel/awx/main/managers.py", line 108, in me
tools_awx_1_1 |     raise RuntimeError("No instance found with the current cluster host id")
tools_awx_1_1 | RuntimeError: No instance found with the current cluster host id
```

Reply:
💥